component/prometheus/relabel/relabel.go

```diff
@@ -95,19 +98,27 @@ func (c *Component) Update(args component.Arguments) error {
 
 // Receive implements the receiver.Receive func that allows an array of metrics
 // to be passed around.
-// TODO (@tpaschalis) The relabelling process will run _every_ time, for all
-// metrics, resulting in some serious CPU overhead. We should be caching the
-// relabeling results per refID and clearing entries for dropped or stale
-// series. This is a blocker for releasing a production-grade version of the
-// prometheus.relabel component.
 func (c *Component) Receive(ts int64, metricArr []*prometheus.FlowMetric) {
 	c.mut.RLock()
 	defer c.mut.RUnlock()
 
 	relabelledMetrics := make([]*prometheus.FlowMetric, 0)
 	for _, m := range metricArr {
 		// Relabel may return the original flowmetric if no changes applied, nil if everything was removed or an entirely new flowmetric.
-		relabelledFm := m.Relabel(c.mrc...)
+		var relabelledFm *prometheus.FlowMetric
+		if fm, found := c.cache[m.GlobalRefID()]; found {
+			relabelledFm = fm
+		} else {
+			relabelledFm = m.Relabel(c.mrc...)
+			c.cache[m.GlobalRefID()] = relabelledFm
+		}
+
+		// If stale remove from the cache, the reason we don't exit early is so the stale value can propagate.
+		// TODO: (@mattdurham) This caching can leak and likely needs a timed eviction at some point, but this is simple.
```

Comment:
Yeah, leaking will _definitely_ happen if a component is removed, since stale markers don't get sent on component shutdown (and Prometheus-team doesn't recommend it working that way). 

This is probably fine for now. As you mentioned, improvements to the global cache to assist in garbage collecting will help with this. 

(Please do create a new issue for us after this is merged to track the improvements so we don't forget that we need to do this though)

Reply:
Created issue #2383 to handle the longer term thoughts.